### PR TITLE
add: preserve dependency order and keep entries in single line

### PR DESCRIPTION
**Pull request by SunPodder**
- Preserve table orders when adding dependencies to the manifest
- Keep dependency entries as a single line table

### CODE REVIEW

File: src/Cmd/Add.cc [hunk 1]
```diff
@@ -65,27 +67,51 @@ handleDependency(
...
-  if (dep.find("://") == std::string_view::npos) {
-    // check if atleast in "user/repo" format
-    if (dep.find('/') == std::string_view::npos) {
-      logger::error("Invalid dependency: " + std::string(dep));
-      return "";
-    }
+  std::regex user_repo_regex(R"(^([^/]+)/([^/]+)$)");
+  std::regex url_regex(
+      R"((?:https?://)?(?:www\.)?(?:github|gitlab)\.com/([^/]+/[^/]+)(?:\.git)?)"
+  );
```
Comment (ken-matsui) [inline src/Cmd/Add.cc:73]: Thank you for your contribution! For consistency and performance reasons, I generally avoid regex in this codebase, as custom parsers tend to offer better performance and readability. Could you please replace the regex with a custom parser here?

Comment (SunPodder) [reply src/Cmd/Add.cc]: Existing implementation works fine in that case, reverted the commit.

**Review by ken-matsui — COMMENTED** [review]
Please update the PR title, what are you enhancing of subcmd Add?

**Comment by SunPodder** [comment]
> Please update the PR title, what are you enhancing of subcmd Add?

done. But what's with the MacOS build? I didn't change any dependency versioning. :\

**Comment by ken-matsui** [comment]
> > Please update the PR title, what are you enhancing of subcmd Add?
> 
> done. 

Hmm, still vague, what did you fix of manifest serialization?

> But what's with the MacOS build? I didn't change any dependency versioning. :\

Please see https://github.com/poac-dev/poac/pull/1032, not sure my code is compatible with tbb 2022 though.

**Comment by ken-matsui** [comment]
> Keep dependency entries as a single line table

Can you please clarify what this means?

**Comment by SunPodder** [comment]
> Can you please clarify what this means?

When you add a dependency, before it was added like this.
```toml
[dependencies.x]
git = "...."

[dependencies.y]
git = "...."
```
Now:
```toml
[dependencies]
x = { git = "...." }
y = { git = "...." }
```

**Comment by ken-matsui** [comment]
This sounds great, thanks. Would you mind adding a comment to the code as well just like you sent?

**Comment by ken-matsui** [comment]
Please fix the CI error, then LGTM. Thanks for your work!

File: src/Cmd/Add.cc [hunk 2]
```diff
@@ -124,8 +127,15 @@ addDependencyToManifest(
...
-  auto data = toml::parse(getManifestPath());
-  auto& deps = toml::find<toml::table>(data, "dependencies");
+  // parse with ordered_type_config to keep the order of the tables
+  auto data = toml::parse<toml::ordered_type_config>(getManifestPath());
+
+  // Ensure the "dependencies" key exists and is a table for correct
+  // serialization
+  if (!data["dependencies"].is_table()) {
+    data["dependencies"] = toml::table{};
+  }
```
Comment (ken-matsui) [inline src/Cmd/Add.cc:137]: If this isn't a table, shouldn't we throw an exception? Overwriting the existing, incorrect data without notifying the user is kind of aggressive, I think.

File: src/Cmd/Add.cc [hunk 3]
```diff
@@ -124,8 +127,15 @@ addDependencyToManifest(
     }
   }
 
-  auto data = toml::parse(getManifestPath());
-  auto& deps = toml::find<toml::table>(data, "dependencies");
+  // parse with ordered_type_config to keep the order of the tables
```
Comment (ken-matsui) [inline src/Cmd/Add.cc:130]:
```suggestion
  // Keep the order of the tables.
```

File: src/Cmd/Add.cc [hunk 4]
```diff
@@ -5,6 +5,7 @@
 #include "../Rustify.hpp"
 #include "Common.hpp"
 
+#include <cstddef>
```
Comment (ken-matsui) [inline src/Cmd/Add.cc:8]: Is this needed?

File: src/Cmd/Add.cc [hunk 5]
```diff
@@ -103,6 +104,8 @@ addDependencyToManifest(
     std::string& rev, std::string& branch
 ) {
   toml::value depData = toml::table{};
+  // Set the formatting for the dependency data table to be on a single line
```
Comment (ken-matsui) [inline src/Cmd/Add.cc:107]: This isn't clear. Let's add an example in this comment.

**Review by ken-matsui — COMMENTED** [review]
Let's address more details. Can you also add a period at the end of comments?

File: src/Cmd/Add.cc [hunk 6]
```diff
@@ -161,7 +177,7 @@ addMain(const std::span<const std::string_view> args) {
   std::unordered_set<std::string_view> newDeps = {};
 
   bool isSystemDependency = false;
-  std::string version; // Only used with system-dependencies
+  std::string version; // Only used with system-dependencies.
```
Comment (ken-matsui) [inline src/Cmd/Add.cc:180]: This comment is not a sentence, adjective, so no period makes more sense.

File: src/Cmd/Add.cc [hunk 7]
```diff
@@ -124,8 +127,21 @@ addDependencyToManifest(
...
+  auto data = toml::parse<toml::ordered_type_config>(getManifestPath());
+
+  // Ensure the "dependencies" key exists and is a table for correct
+  // serialization.
+  if (data["dependencies"].is_empty()) {
+    data["dependencies"] = toml::table{};
+  }
+  if (!data["dependencies"].is_table()) {
+    logger::error(
+        "The `dependencies` key in the poac.toml is not the correct type"
```
Comment (ken-matsui) [inline src/Cmd/Add.cc:140]: This error message is ambiguous. We should show what it should be:
```suggestion
        "The `dependencies` key in the poac.toml should be a table type"
```

Also, use [`make_error_info`](https://github.com/ToruNiina/toml11?tab=readme-ov-file#error-messages) to provide a clearer error message.

File: src/Cmd/Add.cc [hunk 8]
```diff
@@ -66,7 +66,7 @@ handleDependency(
 static std::string
 getDependencyGitUrl(const std::string_view dep) {
   if (dep.find("://") == std::string_view::npos) {
-    // check if atleast in "user/repo" format
+    // check if atleast in "user/repo" format.
```
Comment (ken-matsui) [inline src/Cmd/Add.cc:69]:
```suggestion
    // Check if at least in "user/repo" format.
```

File: src/Cmd/Add.cc [hunk 9]
```diff
@@ -88,7 +88,7 @@ getDependencyName(const std::string_view dep) {
     );
   }
 
-  // Remove trailing '.git' if it exists
+  // Remove trailing '.git' if it exists.
   if (name.ends_with(".git")) {
     name = name.substr(0, name.size() - 4);
```
Comment (ken-matsui) [inline src/Cmd/Add.cc:93]:
```suggestion
    name = name.substr(0, name.size() - ".git"sv.size());
```

File: src/Cmd/Add.cc [hunk 10]
```diff
@@ -124,8 +127,14 @@ addDependencyToManifest(
...
-  auto data = toml::parse(getManifestPath());
-  auto& deps = toml::find<toml::table>(data, "dependencies");
+  // Keep the order of the tables.
+  auto data = toml::parse<toml::ordered_type_config>(getManifestPath());
+
+  // Check if the dependencies table exists, if not create it.
+  if (data["dependencies"].is_empty()) {
+    data["dependencies"] = toml::table{};
+  }
+  auto& deps = data["dependencies"];
```
Comment (ken-matsui) [inline src/Cmd/Add.cc:137]: We should handle the case where `data["dependencies"]` is not a table, shouldn't we?

Comment (ken-matsui) [reply src/Cmd/Add.cc:137]: Great, thanks for your clarification.

**Review by ken-matsui — COMMENTED** [review]
LGTM, thanks!

**Pull request merged**